File: python-sdk/examples/helpers/datapi.py
"""DatAPI client used by the production streaming example.

Mirrors ``rust-sdk/examples/helpers/datapi.rs``: a tiny HTTP wrapper over
``GET /v1/prices?ids=...`` that returns USD prices for the given tokens.
"""

from dataclasses import dataclass
from typing import Dict, Iterable

import requests


@dataclass
class TokenPriceData:
    """Single token's price snapshot returned by DatAPI.

    Mirrors the Rust ``TokenPriceData`` struct.
    """

    usd_price: float
    block_id: int
    decimals: int
    price_change24h: float


# Mirrors Rust's ``DatapiResponse = HashMap<String, TokenPriceData>``.
DatapiResponse = Dict[str, TokenPriceData]


class DatapiClient:
    """Tiny synchronous HTTP client for the Jupiter DatAPI."""

    def __init__(self, host: str):
        self._host = host.rstrip("/")
        self._session = requests.Session()

    def fetch_prices(self, token_ids: Iterable[str]) -> DatapiResponse:
        """Fetch prices for a list of token mints."""
        ids = ",".join(token_ids)
        url = f"{self._host}/v1/prices?ids={ids}"

        response = self._session.get(
            url, headers={"accept": "application/json"}
        )
        if response.status_code != 200:
            raise RuntimeError(
                f"HTTP error {response.status_code}: {response.text}"
            )

        body = response.json()
        return {
            token: TokenPriceData(
                usd_price=float(data.get("usdPrice", 0.0)),
                block_id=int(data.get("blockId", 0)),
                decimals=int(data.get("decimals", 0)),
                price_change24h=float(data.get("priceChange24h", 0.0)),
            )
            for token, data in body.items()
        }

    def fetch_price(self, token_id: str) -> DatapiResponse:
        """Fetch the price for a single token."""
        return self.fetch_prices([token_id])


__all__ = ["DatapiClient", "DatapiResponse", "TokenPriceData"]

File: rust-sdk/examples/helpers/datapi.rs
use serde::Deserialize;
use std::collections::HashMap;
use std::error::Error;

#[derive(Default, Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TokenPriceData {
    pub usd_price: f64,
    pub block_id: u64,
    pub decimals: u8,
    pub price_change24h: f64,
}

pub type DatapiResponse = HashMap<String, TokenPriceData>;

pub struct DatapiClient {
    client: reqwest::Client,
    host: String,
}

impl DatapiClient {
    pub fn new(url: &str) -> Self {
        Self {
            client: reqwest::Client::new(),
            host: url.to_string(),
        }
    }

    pub async fn fetch_prices(
        &self,
        token_ids: &[String],
    ) -> Result<DatapiResponse, Box<dyn Error + Send + Sync>> {
        let ids = token_ids.join(",");
        let url = format!("{}/v1/prices?ids={}", self.host, ids);

        tracing::debug!("fetching prices for tokens {:?}", token_ids);

        let response = self
            .client
            .get(&url)
            .header("accept", "application/json")
            .send()
            .await?;

        if !response.status().is_success() {
            return Err(format!("HTTP error: {}", response.status()).into());
        }

        let price_response: DatapiResponse = response.json().await?;
        Ok(price_response)
    }
}
